### PR TITLE
[aggregator] Add compatibility for rollup rules with timed metrics

### PR DESCRIPTION
**Pull request by robskillington**
**What this PR does / why we need it**:

Adds compatibility for rollup rules to the aggregator with timed metric types.

**Special notes for your reviewer**:

**Does this PR introduce a user-facing and/or backwards incompatible change?**:

```release-note
NONE
```

**Does this PR require updating code package or user-facing documentation?**:

```documentation-note
NONE
```

### CODE REVIEW

File: src/aggregator/aggregation/options.go [hunk 1]
```diff
@@ -54,7 +60,21 @@ type GaugeMetrics struct {
...
+
+func newCounterMetrics(scope tally.Scope) CounterMetrics {
+	return CounterMetrics{
+		valuesOutOfOrder: scope.Counter("values-out-of-order"),
+	}
+}
+
+// IncValuesOutOfOrder increments value or if not initialized is a no-op.
+func (m CounterMetrics) IncValuesOutOfOrder() {
+	if m.valuesOutOfOrder != nil {
```
Comment (schallert) [inline src/aggregator/aggregation/options.go:76]: Do we expect to initialize a `CounterMetrics` without the constructor or is this just to be paranoid?

File: src/aggregator/aggregator/counter_elem_gen.go [hunk 2]
```diff
@@ -412,28 +412,60 @@ func (e *CounterElem) processValueWithAggregationLock(
...
+				}
+
+				res := binaryOp.Evaluate(prev, curr)
+
 				// NB: we only need to record the value needed for derivative transformations.
 				// We currently only support first-order derivative transformations so we only
 				// need to keep one value. In the future if we need to support higher-order
 				// derivative transformations, we need to store an array of values here.
-				e.lastConsumedValues[aggTypeIdx] = value
+				if !math.IsNaN(curr.Value) {
```
Comment (schallert) [inline src/aggregator/aggregator/counter_elem_gen.go:456]: Discussed offline why we bother passing evaluating the op if incoming sample is NaN but there may be transformations (such as count # samples) that still care.

File: src/aggregator/aggregator/entry.go [hunk 3]
```diff
@@ -651,6 +667,71 @@ func (e *Entry) addTimed(
...
+
+		// If the metadata indicates the (rollup) metric has been tombstoned, the metric is
+		// not ingested for aggregation. However, we do not update the policies asssociated
+		// with this entry and mark it tombstoned because there may be a different raw metric
+		// generating this same (rollup) metric that is actively emitting, meaning this entry
+		// may still be very much alive.
+		if sm.Tombstoned {
+			e.RUnlock()
+			timeLock.RUnlock()
+			// e.metrics.timed.tombstonedMetadata.Inc(1)
```
Comment (schallert) [inline src/aggregator/aggregator/entry.go:689]: Undo comment?

File: src/aggregator/aggregator/entry.go [hunk 4]
```diff
@@ -651,6 +667,71 @@ func (e *Entry) addTimed(
...
+			return nil
+		}
+
+		// It is expected that there is at least one pipeline in the metadata.
+		if len(sm.Pipelines) == 0 {
+			e.RUnlock()
+			timeLock.RUnlock()
+			e.metrics.untimed.noPipelinesInMetadata.Inc(1)
```
Comment (schallert) [inline src/aggregator/aggregator/entry.go:697]: `s/untimed/timed/`?

File: src/metrics/transformation/binary.go [hunk 5]
```diff
@@ -45,3 +57,26 @@ func perSecond(prev, curr Datapoint) Datapoint {
...
+// * It assumes the timestamps are monotonically increasing, and values are non-decreasing.
+//   If the value decreases, this is treated as a counter reset and the value is returned.
+func increase(prev, curr Datapoint) Datapoint {
+	if prev.TimeNanos >= curr.TimeNanos || math.IsNaN(prev.Value) || math.IsNaN(curr.Value) {
+		return emptyDatapoint
+	}
+	diff := curr.Value - prev.Value
+	if diff < 0 {
+		// // Treated as a counter reset, take the curr value as counted elements.
+		// return Datapoint{TimeNanos: curr.TimeNanos, Value: curr.Value}
```
Comment (schallert) [inline src/metrics/transformation/binary.go:78]: Stale comment

File: src/aggregator/aggregation/options.go (hunk 1 above)
Comment (robskillington) [reply src/aggregator/aggregation/options.go:76]: Bit of both, I believe some integration tests do not initialize it and also this is for paranoia.

File: src/aggregator/aggregator/counter_elem_gen.go (hunk 2 above)
Comment (robskillington) [reply src/aggregator/aggregator/counter_elem_gen.go:456]: 👍

File: src/aggregator/aggregator/entry.go (hunk 3 above)
Comment (robskillington) [reply src/aggregator/aggregator/entry.go:689]: Ah good call.

File: src/aggregator/aggregator/entry.go (hunk 4 above)
Comment (robskillington) [reply src/aggregator/aggregator/entry.go:697]: Yup, ty good call.

**Pull request merged**